File: cpp/543. Diameter of Binary Tree.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int ans = 0;
    int dfs(TreeNode *root) { // return the depth of root
        int d = 0, dd = 0;
        if(root->left) {
            int d1 = dfs(root->left);
            dd += d1+1;
            d = max(d, d1+1);
        }
        if(root->right) {
            int d2 = dfs(root->right);
            dd += d2+1;
            d = max(d, d2+1);
        }
        ans = max(ans, dd);
        return d;
    }
    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/543-Diameter-of-Binary-Tree.py
from bintree_helper import *

class Solution:
    def diameterOfBinaryTree(self, root) -> int:
        "DFS, Time: O(n), Space: O(n)"
        if root is None: return 0
        result = 0
        def depth(node):
            "return the max depth from node to leaf"
            nonlocal result
            ld, rd = 0, 0
            if node.left:
                ld = depth(node.left) + 1
            if node.right:
                rd = depth(node.right) + 1
            result = max(result, ld+rd)
            return max(ld, rd)
        depth(root)
        return result

root = "[1,2,3,4,5,null,null]"
root = deserialize(root)
result = Solution().diameterOfBinaryTree(root)
assert result == 3

root = "[4,-7,-3,null,null,-9,-3,9,-7,-4,null,6,null,-6,-6,null,null,0,6,5,null,9,null,null,-1,-4,null,null,null,-2]"
root = deserialize(root)
result = Solution().diameterOfBinaryTree(root)
assert result == 8